src/test/java/com/atlassian/db/replica/api/TestDualConnection.java

```diff
@@ -1653,6 +1655,37 @@ public void shouldKeepUsingReplicaAfterSettingTimeout() throws SQLException {
             .containsOnly(REPLICA);
     }
 
+    @Test
+    public void shouldKeepStatementTimeoutWhenSwitchingConnections() throws SQLException {
+        final ConnectionProviderMock connectionProvider = new ConnectionProviderMock();
+        final Connection connection = DualConnection.builder(
+            connectionProvider,
+            permanentConsistency().build()
+        ).build();
+
+        try (Statement statement = connection.createStatement()) {
+            statement.execute("set statement_timeout to 30000");
+        }
+
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+        connection.prepareStatement(SIMPLE_QUERY).executeQuery();
+        final List<Statement> replicaStatements = connectionProvider.getStatements();
+        connection.prepareStatement(SELECT_FOR_UPDATE).executeQuery();
+        connection.prepareStatement(SELECT_FOR_UPDATE).executeQuery();
+
+        final List<Statement> allStatements = connectionProvider.getStatements();
+        final List<Statement> mainConnectionStatements = allStatements
+            .stream()
+            .filter(statement -> !replicaStatements.contains(statement))
+            .collect(Collectors.toList());
+        assertThat(connectionProvider.getProvidedConnectionTypes())
+            .containsExactly(REPLICA, MAIN);
+        assertThat(replicaStatements).hasSize(1);
+        verify(replicaStatements.get(0)).execute("set statement_timeout to 30000");
+        assertThat(mainConnectionStatements).hasSize(1);
+        verify(mainConnectionStatements.get(0)).execute("set statement_timeout to 30000");
```

Comment:
is there option to have different `set statement_timeout to 30000` only on MAIN ?

Reply:
`set` changes configuration for a current connection. From the perspective of DualConnection's user, there's no replica/main. There's only DualConnection. The user can't use different settings for main and replica as the internal connections are transparent and there's no dedicated API to set one or another.

Reply:
Correct, I was wondering about some edge case, but it doesn't make senses as
> internal connections are transparent

👍